File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        n = len(prices)
        dp = [0] * n

        min_buy = prices[0]

        for i in range(1, n):
            dp[i] = max(dp[i - 1], prices[i] - min_buy)
            min_buy = min(min_buy, prices[i])

        return dp[-1]





        
       

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int minBuy = prices[0];

        for (int price : prices) {
            if (price < minBuy) minBuy = price;                 // best (cheapest) buy so far
            else maxProfit = Math.max(maxProfit, price - minBuy); // best profit if sold today
        }
        return maxProfit;
    }
}
